eval/src/vespa/eval/tensor/serialization/dense_binary_format.cpp

```diff
@@ -54,16 +74,75 @@ DenseBinaryFormat::deserialize(nbostream &stream)
         dimensions.emplace_back(dimensionName, dimensionSize);
         cellsSize *= dimensionSize;
     }
-    cells.reserve(cellsSize);
-    double cellValue = 0.0;
+    return cellsSize;
+}
+
+SerializationFormat
+decodeCellType(nbostream & stream, EncodeType encodeType) {
+    if (encodeType != EncodeType::DOUBLE_IS_DEFAULT) {
+        uint32_t serializedType = stream.getInt1_4Bytes();
+        switch (serializedType) {
+            case DOUBLE_VALUE_TYPE:
+                return SerializationFormat::DOUBLE;
+            case FLOAT_VALUE_TYPE:
+                return  SerializationFormat::FLOAT;
+            default:
+                throw IllegalArgumentException(make_string("Received unknown tensor value type = %u. Only 0(double), or 1(float) are legal.", serializedType));
+        }
+    } else {
+        return SerializationFormat::DOUBLE;
+    }
+}
+
+template<typename T>
+void
+decodeCells(nbostream &stream, size_t cellsSize, DenseTensor::Cells & cells) {
+    T cellValue = 0.0;
     for (size_t i = 0; i < cellsSize; ++i) {
         stream >> cellValue;
         cells.emplace_back(cellValue);
     }
-    return std::make_unique<DenseTensor>(makeValueType(std::move(dimensions)),
-                                         std::move(cells));
 }
 
+}
+
+void
+DenseBinaryFormat::serialize(nbostream &stream, const DenseTensorView &tensor)
+{
+    const eval::ValueType & type = tensor.fast_type();
+    encodeValueType(stream, tensor.serializeAs(), _encodeType);
+    size_t cellsSize = encodeDimensions(stream, type);
+
+    DenseTensorView::CellsRef cells = tensor.cellsRef();
+    assert(cells.size() == cellsSize);
+    switch (tensor.serializeAs()) {
+        case SerializationFormat::DOUBLE:
+            encodeCells<double>(stream, cells);
+            break;
+        case SerializationFormat::FLOAT:
+            encodeCells<float>(stream, cells);
```

Comment:
this will break tensor evaluation for cases not directly supported by dense tensors (mixed tensors etc...)

Reply:
Reverted as discussed